src/text-input/index.tsx

```diff
@@ -252,91 +235,110 @@ export class TextInput extends ThemedMixin(FocusMixin(WidgetBase))<TextInputProp
 
 		const computedHelperText = (valid === false && message) || helperText;
 
-		return v(
-			'div',
-			{
-				key: 'root',
-				classes: this.theme(this.getRootClasses()),
-				role: 'presentation'
-			},
-			[
-				label &&
-					w(
-						Label,
-						{
-							theme,
-							classes,
-							disabled,
-							valid,
-							focused: focus.containsFocus,
-							readOnly,
-							required,
-							hidden: labelHidden,
-							forId: widgetId
-						},
-						[label]
-					),
-				v(
-					'div',
-					{
-						key: 'inputWrapper',
-						classes: this.theme(css.inputWrapper),
-						role: 'presentation'
-					},
-					[
-						leading &&
-							v('span', { key: 'leading', classes: this.theme(css.leading) }, [
-								leading()
-							]),
-						v('input', {
-							...formatAriaProperties(aria),
-							'aria-invalid': valid === false ? 'true' : null,
-							autocomplete: formatAutocomplete(autocomplete),
-							classes: this.theme(css.input),
-							disabled,
-							id: widgetId,
-							focus: this.shouldFocus,
-							key: 'input',
-							max,
-							maxlength: maxLength ? `${maxLength}` : null,
-							min,
-							minlength: minLength ? `${minLength}` : null,
-							name,
-							pattern,
-							placeholder,
-							readOnly,
-							'aria-readonly': readOnly ? 'true' : null,
-							required,
-							step,
-							type,
-							value,
-							onblur: () => {
+		return (
+			<div key="root" classes={this.theme(css.root)} role="presentation">
+				<div
+					key="wrapper"
+					classes={this.theme([
+						css.wrapper,
+						disabled ? css.disabled : null,
+						focus.containsFocus ? css.focused : null,
+						valid === false ? css.invalid : null,
+						valid === true ? css.valid : null,
+						readOnly ? css.readonly : null,
+						required ? css.required : null,
+						leading ? css.hasLeading : null,
+						trailing ? css.hasTrailing : null,
+						!label ? css.noLabel : null
+					])}
+					role="presentation"
+				>
+					{label && (
+						<Label
+							theme={theme}
+							disabled={disabled}
+							valid={valid}
+							focused={focus.containsFocus}
+							readOnly={readOnly}
+							required={required}
+							hidden={labelHidden}
+							forId={widgetId}
+							classes={
```

Comment:
this should really be done via `theme` and `theme.compose` with a prefix of `label`.